tensorboard/backend/http_util.py

```diff
@@ -119,18 +119,24 @@ def Respond(request,
   content = tf.compat.as_bytes(content, charset)
   if textual and not charset_match and mimetype not in _JSON_MIMETYPES:
     content_type += '; charset=' + charset
-  if (not content_encoding and textual and
-      _ALLOWS_GZIP_PATTERN.search(request.headers.get('Accept-Encoding', ''))):
+  gzip_accepted = _ALLOWS_GZIP_PATTERN.search(
+      request.headers.get('Accept-Encoding', ''))
+  # Automatically gzip uncompressed text data if accepted.
+  if textual and not content_encoding and gzip_accepted:
+    orig_len = len(content)
     out = six.BytesIO()
-    f = gzip.GzipFile(fileobj=out, mode='wb', compresslevel=3)
-    f.write(content)
-    f.close()
+    # Set mtime to zero to make payload for a given input deterministic.
+    with gzip.GzipFile(fileobj=out, mode='wb', compresslevel=3, mtime=0) as f:
+      f.write(content)
     content = out.getvalue()
     content_encoding = 'gzip'
-  if request.method == 'HEAD':
-    content = ''
-  headers = []
+  # Automatically unzip precompressed data if not accepted.
+  if content_encoding == 'gzip' and not gzip_accepted:
+    with gzip.GzipFile(fileobj=six.BytesIO(content), mode='rb') as f:
```

Comment:
[optional and probably not worth doing]

If you care about shaving tens of milliseconds or so for the 0.001% of requests that don't accept gzip, you should be able to say: `content = gzip.GzipFile(...); [...]; return wrappers.Response(response=content, direct_passthrough=True)`. Please note that in order for this to work with HTTP/1.1 you would need to get rid of the content-length and make sure Werkzeug is using chunked transfer encoding under the hood. Or say `Connection: close`.

Reply:
PTAL - I actually decided to implement this, thanks for the tip!  Now it does a streaming gunzip, and indeed that eliminates the 20 millisecond delay in TTFB when doing on-demand decompression - it's <5 ms to the first byte returned now when decompressing on demand :)

The Content-Length hassles would have been a dealbreaker IMO, but as it so happens gzip files encode the length of the original uncompressed content in their last four bytes (well, strictly speaking they record the lower 32 bits of that length, but I think we can rely on not having pre-compressed blobs larger than 2^32 bytes = 4GB).  So we can extract those and use them to send the Content-Length header before we've decompressed a single byte.

Reply:
That's such a great detail you uncovered!